pkg/network/bridge_vlan_controller.go

```diff
@@ -67,7 +79,81 @@ func (c *BridgeVLANController) OnRemove(key string, setting *harvesterv1.Setting
 		return setting, nil
 	}
 
-	// TODO, bridge VLAN setting is not suppose to be removed, therefore we can add a backup setting on delete?
+	networkSetting, err := decodeNetworkSettings(setting.Value)
+	if err != nil {
+		return nil, fmt.Errorf("decode failed, error: %w, value: %s", err, setting.Value)
+	}
+
+	nic, err := bridge.GetLink(networkSetting.NIC)
+	if err != nil {
+		klog.Error(err)
+		return setting, nil
+	}
+
+	// if return address failed, log and continue
+	if err := c.bridge.ReturnAddr(nic); err != nil {
+		klog.Errorf("return address failed, error: %v, NIC: %s", err, networkSetting.NIC)
+	}
+
+	if err := c.bridge.Delete(); err != nil {
+		klog.Errorf("delete bridge failed, error: %v, bridge: %+v", err, c.bridge)
+	}
 
 	return setting, nil
 }
+
+func (c *BridgeVLANController) configBridgeNetwork(setting *NetworkSetting) error {
+	nic, err := bridge.GetLink(setting.NIC)
+	if err != nil {
+		return err
+	}
+	configuredNIC, err := bridge.GetLink(setting.ConfiguredNIC)
+	if err != nil {
+		return err
+	}
+
+	// ensure bridge existed
+	if err := c.bridge.Ensure(); err != nil {
+		return fmt.Errorf("ensure bridge failed, error: %w, bridge: %+v", err, c.bridge)
+	}
+
+	// nic is nil if setting.NIC is an empty string
+	// return addresses to configured nic if nic is empty
+	if nic == nil {
+		if err := c.bridge.ReturnAddr(configuredNIC); err != nil {
+			klog.Errorf("return address failed, error: %v, NIC: %s", err, setting.ConfiguredNIC)
+		}
+		return nil
+	}
+
+	// Check whether the master of nic is bridge
+	// - Yes，do nothing
+	// - No, make bridge return addresses to configured nic, lend addresses to bridge, return.
+	if nic.Attrs().MasterIndex != c.bridge.Index {
+		if err := c.bridge.ReturnAddr(configuredNIC); err != nil {
+			klog.Errorf("Return address failed, error: %v, NIC: %s", err, setting.ConfiguredNIC)
```

Comment:
Why do we proceed when ReturnAddr fails? (including the L124)

Reply:
> Why do we proceed when ReturnAddr fails? (including the L124)

In some case, ReturnAddr will fail, for example, previous configured nic is deleted manually. But I think the bridge configuration process needs to go on.